sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/CatalogManagerSchema.java

```diff
@@ -0,0 +1,284 @@
+          RESOURCE.internal(String.format("Cannot use catalog: '%s' not found.", name)));
+    }
+
+    if (catalogManager.currentCatalog().name().equals(name)) {
+      LOG.info("Catalog '{}' is already in use.", name);
+      return;
+    }
+
+    catalogManager.useCatalog(name);
+    LOG.info("Switched to catalog '{}' (type: {})", name, catalogManager.currentCatalog().type());
+  }
+
+  public void dropCatalog(SqlIdentifier identifier, boolean ifExists) {
+    String name = SqlDdlNodes.name(identifier);
+    if (catalogManager.getCatalog(name) == null) {
+      if (!ifExists) {
+        throw SqlUtil.newContextException(
+            identifier.getParserPosition(),
+            RESOURCE.internal(String.format("Cannot drop catalog: '%s' not found.", name)));
+      }
+      LOG.info("Ignoring 'DROP CATALOG` call for non-existent catalog: {}", name);
+      return;
+    }
+
+    if (catalogManager.currentCatalog().name().equals(name)) {
+      throw SqlUtil.newContextException(
+          identifier.getParserPosition(),
+          RESOURCE.internal(
+              String.format(
+                  "Unable to drop active catalog '%s'. Please switch to another catalog first.",
+                  name)));
+    }
+
+    catalogManager.dropCatalog(name);
+    LOG.info("Successfully dropped catalog '{}'", name);
+    catalogSubSchemas.remove(name);
+  }
+
+  // A BeamCalciteSchema may be used to interact with multiple TableProviders.
+  // If such a TableProvider is not registered in the BeamCalciteSchema, this method
+  // will attempt to do so.
+  public void maybeRegisterProvider(TableName path, String type) {
+    CatalogSchema catalogSchema = getCatalogSchema(path);
+    BeamCalciteSchema beamCalciteSchema = catalogSchema.getDatabaseSchema(path);
+
+    if (beamCalciteSchema.getTableProvider() instanceof MetaStore) {
+      MetaStore metaStore = (MetaStore) beamCalciteSchema.getTableProvider();
+      if (metaStore.tableProviders().containsKey(type)) {
+        return;
+      }
+
+      // Start with the narrowest scope.
+      // Attempt to fetch provider from Catalog first, then CatalogManager.
+      @Nullable TableProvider provider = catalogSchema.getCatalog().tableProviders().get(type);
+      if (provider == null) {
+        provider = catalogManager.tableProviders().get(type);
+      }
+      // register provider
+      if (provider != null) {
+        metaStore.registerProvider(provider);
+      }
+    }
+  }
+
+  @Override
+  public @Nullable Table getTable(String table) {
+    @Nullable
+    CatalogSchema catalogSchema = catalogSubSchemas.get(catalogManager.currentCatalog().name());
+    return catalogSchema != null ? catalogSchema.getTable(table) : null;
+  }
+
+  @Override
+  public Set<String> getTableNames() {
+    ImmutableSet.Builder<String> names = ImmutableSet.builder();
+    // TODO: this might be a heavy operation
+    for (CatalogSchema catalogSchema : catalogSubSchemas.values()) {
```

Comment:
@ahmedabu98 I believe this is very heavy method, can we only list tables within the currently active catalog and database ?